Source/Notifications/Push notifications/Notification Types/Content/ZMLocalNotification+Events.swift

```diff
@@ -269,3 +279,101 @@ private class NewUserEventNotificationBuilder: EventNotificationBuilder {
+        if case .ephemeral? = message.content {
+            return LocalNotificationType.message(.hidden)
+        }
+        return LocalNotificationDispatcher.shouldHideNotificationContent(moc: self.moc)
+        ? LocalNotificationType.message(.hidden)
+        : LocalNotificationType.message(contentType)
+    }
+
+    override func shouldCreateNotification() -> Bool {
+        guard let conversation = conversation,
+            let senderUUID = event.senderUUID(),
+            !conversation.isMessageSilenced(message, senderID: senderUUID) else {
+                Logging.push.safePublic("Not creating local notification for message with nonce = \(event.messageNonce) because conversation is silenced")
+                return false
+        }
+        
+        if let timeStamp = event.timeStamp(),
+            let lastRead = conversation.lastReadServerTimeStamp,
+            lastRead.compare(timeStamp) != .orderedAscending
+        {
+            return false
+        }
+        return true
+    }
+}
+
+// MARK: - System Message
+
+private class NewSystemMessageNotificationBuilder : EventNotificationBuilder {
+    let contentType: LocalNotificationContentType
+    
+    override init?(event: ZMUpdateEvent, conversation: ZMConversation?, managedObjectContext: NSManagedObjectContext) {
+        guard let contentType = LocalNotificationContentType.typeForMessage(event, conversation: conversation, in: managedObjectContext)
+            else {
+                return nil
+        }
+        
+        self.contentType = contentType
+        super.init(event: event, conversation: conversation, managedObjectContext: managedObjectContext)
+    }
+    
+    override func titleText() -> String? {
+        return notificationType.titleText(selfUser: ZMUser.selfUser(in: moc), conversation: conversation)
+    }
+    
+    override func bodyText() -> String {
+        return notificationType.messageBodyText(sender: sender, conversation: conversation).trimmingCharacters(in: .whitespaces)
+    }
+    
+    override var notificationType: LocalNotificationType {
+        return LocalNotificationType.message(contentType)
+    }
+    
+    override func shouldCreateNotification() -> Bool {
+        // we don't want to create notifications when other people join or leave conversation
+        guard let dataPayload = (event.payload as NSDictionary).dictionary(forKey: "data"),
+            let userIds = dataPayload["user_ids"] as? [String] else {
+                return false
+        }
+        let users = userIds.compactMap({ UUID.init(uuidString: $0)}).compactMap({ZMUser(remoteID: $0, createIfNeeded: true, in: moc)})
+        let forSelf = users.count == 1 && users.first!.isSelfUser
+
+        switch contentType {
+        case .participantsAdded, .participantsRemoved:
+            return forSelf ? super.shouldCreateNotification() : false
+        default:
+            return super.shouldCreateNotification()
+        }
```

Comment:
I now realise that my suggestions was not maybe so great since it complicates the flow to call super. What about something like this?

```suggestion
        switch contentType {
        case .participantsAdded, .participantsRemoved where concernsSelfUser == false
            return false
        default:
            break
        }
        
        return super.shouldCreateNotification()
```

Reply:
Done